core/monitor/monitor.go

```diff
@@ -92,49 +93,104 @@ func New(mclient pb.MutationServiceClient,
 	}, nil
 }
 
-// ProcessLoop continuously fetches mutations and processes them.
-func (m *Monitor) ProcessLoop(domainID string, period time.Duration) {
-	mutCli := mutationclient.New(m.mClient, period)
-	responses, errs := mutCli.StartPolling(domainID, 1)
-	for {
+// EpochPair is two adjacent epochs.
+type EpochPair struct {
+	EpochA, EpochB *pb.Epoch
+}
+
+// EpochPairs consumes epochs (0, 1, 2) and produces pairs (0,1), (1,2).
+func EpochPairs(ctx context.Context, epochs <-chan *pb.Epoch, pairs chan<- EpochPair) error {
+	var epochA *pb.Epoch
+	for epoch := range epochs {
+		if epochA == nil {
+			epochA = epoch
+			continue
+		}
+		pair := EpochPair{
+			EpochA: epochA,
+			EpochB: epoch,
+		}
 		select {
-		case mutResp := <-responses:
-			glog.Infof("Received mutations response: %v", mutResp.Epoch)
-			if err := m.Process(mutResp); err != nil {
-				glog.Infof("Error processing mutations response: %v", err)
-			}
-		case err := <-errs:
-			// this is OK if there were no mutations in  between:
-			// TODO(ismail): handle the case when the known maxDuration has
-			// passed and no epoch was issued?
-			glog.Infof("Could not retrieve mutations API response %v", err)
+		case <-ctx.Done():
+			close(pairs)
+			return ctx.Err()
+		case pairs <- pair:
 		}
+		epochA = epoch
 	}
-
+	return nil
 }
 
-// Process processes a mutation response received from the keytransparency
-// server. Processing includes verifying, signing and storing the resulting
-// monitoring response.
-func (m *Monitor) Process(resp *pb.GetMutationsResponse) error {
-	var smr *trillian.SignedMapRoot
-	var err error
-	errs := m.VerifyMutationsResponse(resp)
-	if len(errs) == 0 {
-		glog.Infof("Successfully verified mutations response for epoch: %v", resp.Epoch)
-		smr, err = m.signMapRoot(resp)
+// ProcessLoop continuously fetches mutations and processes them.
+func (m *Monitor) ProcessLoop(ctx context.Context, domainID string, period time.Duration) error {
+	mutCli := mutationclient.New(m.mClient, period)
+	cctx, cancel := context.WithCancel(ctx)
+	errc := make(chan error)
+	epochs := make(chan *pb.Epoch)
+	pairs := make(chan EpochPair)
+
+	go func(ctx context.Context, domainID string, epochs chan<- *pb.Epoch) {
```

Comment:
I think it ought to be fine not to pass `domainID` and `epochs` in as arguments, they should be captured from the outer scope?

Reply:
Done.